03/ht/gendos379/01/script.js

```diff
@@ -0,0 +1,118 @@
+var MAX_HISTORY_LENGTH = 5;
+var GOOGLE_API_KEY = 'AIzaSyAPUKxBt2EXY6fBcxAyGIDxnF_0FGGmvgc';
+var $$ = document.querySelector.bind(document);
+
+var lastSearchedCities = JSON.parse(localStorage.getItem('lastSearchedCities')) || [];
+
+function handleCity(city) {
+    res = lastSearchedCities.indexOf(city);
+    if (res >= 0) {
+        lastSearchedCities.splice(res, 1);
+    }
+    lastSearchedCities.unshift(city);
+    lastSearchedCities = lastSearchedCities.slice(0, MAX_HISTORY_LENGTH);
+    localStorage.setItem('lastSearchedCities', JSON.stringify(lastSearchedCities));
+    window.location.hash = city;
+}
+
+function showWeather(city) {
+    fetch(`https://maps.googleapis.com/maps/api/geocode/json?address=${city}&key=${GOOGLE_API_KEY}`)
+    .then(function(response) {
+      return response.json();
+    })
+    .then(function(data) {
+        var placeName = '';
+        for (var i = 0; i < data.results[0].address_components.length; i++) {
+            placeName = data.results[0].address_components[i].short_name + ' ' + placeName;
+        }
+        $$('#weather').innerHTML = placeName + ': <br /><br />';
+        return data.results[0].geometry.location;
+    })
+    .then(function({lat, lng}) {
+        fetch(`https://shrouded-spire-35703.herokuapp.com/forecast/${lat},${lng}?lang=ru&units=si`)
+        .then(function(response) {
+            return response.json();
+        })
+        .then(function(data) {
+            $$('#weather').innerHTML += data.currently.summary + '<br />Температура, C: ' + data.currently.temperature +
+            '<br />Скорость ветра, м/с: ' + data.currently.windSpeed + '<br />Влажность, %: ' + (data.currently.humidity * 100).toFixed();
+        });
+    })
+    .catch(function() {  
+        $$('#weather').innerHTML = 'Введенное место не найдено';
+    });
+}
+
+function showWeatherXHR(city) {
+    var getLatLng = new XMLHttpRequest();
+    getLatLng.open("GET", `https://maps.googleapis.com/maps/api/geocode/json?address=${city}&key=${GOOGLE_API_KEY}`, true);
+    getLatLng.onload = function () {
+        var data = JSON.parse(getLatLng.responseText);
+        if (data.status === 'OK') {
+            var placeName = '';
+            for (var i = 0; i < data.results[0].address_components.length; i++) {
+                placeName = data.results[0].address_components[i].short_name + ' ' + placeName;
+            }
+            $$('#weather').innerHTML = placeName + ': <br /><br />';
+            var {lat, lng} = data.results[0].geometry.location;
+            var getWeather = new XMLHttpRequest();
+            getWeather.open("GET", `https://shrouded-spire-35703.herokuapp.com/forecast/${lat},${lng}?lang=ru&units=si`, true);
+            getWeather.onload = function () {
+                var data = JSON.parse(getWeather.responseText);
```

Comment:
This is not good for me also. This was just a quick solution to make the app work, which is needed to be fixed.. I'll try to reduce this duplication today.